File: DSA-Python/1-tutorials/3-operators.py
# Arithmetic Operators
# Addition (+)
# Subtraction (-)
# Multiplication (*)
# Division (/)
# Floor Division (//)
# Modulo (%)
# Exponentiation (**)
print("4/3 =", 4/3)
print("4//3 =", 4//3)
print("2**3 =", 2**3)
print()



# Assignment operators
# Addition assignment (+=)
# Subtraction assignment (-=)
# Multiplication assignment (*=)
# Division assignment (/=)
# Floor Division assignment (//=)
# Modulo assignment (%=)
# Exponentiation assignment(**=)



# Comparison Operators
# Equal to (==)
# Not Equal to (!=)
# Greater than (>)
# Less than (<)
# Greater than or Equal to (>=)
# Less than or Equal to (<=)



# Logical Operators
# AND (and): Both True, then True. If first condition is False, and does not check the second condition
# OR (or): Both False, then False. If first condition is True, or will not check the second condition
# NOT (not): Negate the operand
print(True and True)
print(False or False)
print(not True)
print()



# Bitwise Operators
# Bitwise AND (&): Both bits at same position 1, then 1
print('8 & 1:', 8 & 1)

# Bitwise OR (|): Both bits at same position 0, then 0
print('8 | 1:', 8 | 1)

# Bitwise XOR (^) or Exclusive OR: Both bits same, then 0
print('8 ^ 0:', 8 ^ 0)

# Bitwise NOT (~) operator flips every bit in the operand’s binary representation. For unsigned integers, this simply inverts the bits. For signed integers (two’s complement), it produces the value -(x + 1).
print('~5:', ~5)
print('~-6:', ~-6)

# Bitwise Left shift (<<) shifts bits to left by n places and add zeroes
# Left shift means multiply by 2^n where n is numver of bits shifted
print('3 << 4:', 3 << 4)
print('-3 << 1:', -3 << 2)

# Bitwise Right shift (>>) operator shifts the bits to the right, and any bits that overflow are discarded
# Right shift means integer floor division by powers of 2:  floor( x/(2^n) ) where n is number of bit to shift
print('3 >> 1:', 3 >> 1)
print()



# Ternary operator
result = 100 if False else 4
print(result)
print()




# Membership Operators
# In (in): Checks if a value exists inside a collection (Array, Set, Hash Map)
str = "Harry"
print("end in friend:", "end" in "friend")

# Not In (not in): Checks if an element is not present in a collection.
print("u not in love:", "u" not in "love")
print()




# Identity Operators
# Is (is): Checks if two variables point to the exact same memory address
# Is Not (is not): Oppsite if 'is' operator
x = 34
y = 34
print(x is y)
print()



# Walrus operator (:=) assigns the value and immediately passes that value forward
ptr = 0
while (ptr := ptr+1) <= 5:
    print(ptr)

File: DSA-C++/1-basics/3-operators.cpp
#include <iostream>
#include <math.h>

int convert_to_base2();
int convert_to_base10();


using namespace std;
int main() {
    // The insertion operator (<<) is used to send data to an output stream, which is often the console but can also be a file or a network connection.
    cout << "Hello" << endl;

    // Extraction operator (>>) is used to read data from an input stream, such as the keyboard, and store it in a variable
    // int t;
    // cin >> t;



    // Arithmetic Operators
    int a = 3;
    int b = 5;
    cout << "a+b = " << a+b << endl;
    cout << "a-b = " << a-b << endl;
    cout << "a*b = " << a*b << endl;
    cout << "a/b = " << a/b << endl;
    cout << "a%b = " << a%b << endl << endl;
    
    

    // Comparison (relational) operators return True or false. 
    // C++ internally converts boolean to integers.
    cout << "a==b " << (a==b) << endl;
    cout << "a!=b " << (a!=b) << endl;
    cout << "a>b " << (a>b) << endl;
    cout << "a<b " << (a<b) << endl;
    cout << "a<=b " << (a<=b) << endl;
    cout << "a>=b " << (a>=b) << endl << endl;
    
 

    // Logical operators always return either 0 or 1
    // && returns true, if both true
    cout << "Logical AND: "<< (5 && 10) << endl;
    // || returns false if both false
    cout << "Logical OR: "<< (5 || 10) << endl;
    // ! Logical NOT negates given operand
    cout << "Logical NOT: "<< (!67) << endl << endl;
    
    


    // Operations involving bitwise operator return any integer.
    // Bitwise AND operator (&) returns result of 1 if both the bits at same position are 1.
    cout << "Bitwise AND: " << (8 & 1) << endl;

    // Bitwise OR (|) returns 0 if both bits at same position are 0
    cout << "Bitwise OR: " << (8 | 0) << endl;

    // Bitwise XOR (^) returns 0 if both operands are same. 
    cout << "Bitwise XOR: " << (6^0) << endl;

    // Bitwise NOT (~) operator flips every bit in the operand’s binary representation. For unsigned integers, this simply inverts the bits. For signed integers (two’s complement), it produces the value -(x + 1).
    cout << "Bitwise NOT: " << (~5) << ", another eg. " << (~-6) << endl;

    // Bitwise Left shift (<<) shifts bits to left by n places and add zeroes
    // Left shift multiplies the number by 2^n where n is numver of bits shifted
    cout << "<< 3 by 4 bits:  " << (3 << 4) << endl; 
    cout << "<< -3 by 1 bits: " << (-3 << 1) << endl;

    // Bitwise Right shift (>>) operator shifts the bits to the right, and any bits that overflow are discarded
    cout << ">> 3 by 1 bits: " << (3 >> 1) << endl << endl;

    

    
    // 5. Assignment Operators
    // These are used as a shorthand for arithmetic operations
    // = Simple assignment
    // +=  Addition assignment
    // -=  Subtraction assignment
    // *=  Multiplication assignment
    // /=  Division assignment
    // %=  Modulus assignment
    // &=  Bitwise AND assignment
    // |=  Bitwise OR assignment
    // ^=  Bitwise XOR assignment
    // <<= Left shift assignment
    // >>= Right shift assignment
    int p = 3;
    p+=5;
    cout << p << endl;


    
    // 6. Increment/Decrement operators increase or decrease value of variable by one
    int x = 6;
    int y = 6;
    cout << "increment and return: " << ++x << endl;    // slightly faster because no copy created
    cout << "return and increment: " << y++ << endl;



    // 7. Conditional (Ternary) Operator
    string result = (1==0) ? "yes": "no";
    cout << result << "\n\n\n";
    return 0;
}